cdap-watchdog/src/main/java/co/cask/cdap/logging/write/FileMetaDataManager.java

```diff
@@ -212,112 +211,92 @@ public TableKey apply(Table table) throws Exception {
     });
   }
 
-  /**
-   * Class which holds table key (row key + column) for log meta
-   */
-  public static final class TableKey {
-    private final String rowKey;
-    private final byte[] startColumn;
-
-    public TableKey(String rowKey, @Nullable byte[] startColumn) {
-      this.rowKey = rowKey;
-      this.startColumn = startColumn;
-    }
-
-    public String getRowKey() {
-      return rowKey;
-    }
-
-    public byte[] getStartColumn() {
-      return startColumn;
+  private byte[] getStopKey(@Nullable TableKey tableKey) {
+    byte[] stopKey;
+    if (tableKey == null) {
+      stopKey = ROW_KEY_PREFIX_END;
+    } else if (Arrays.equals(tableKey.getStopKey(), ROW_KEY_PREFIX_END)) {
+      stopKey = tableKey.getStopKey();
+    } else {
+      stopKey = getRowKey(tableKey.getStopKey());
     }
+    return stopKey;
   }
 
   /**
-   * Deletes meta data until a given time, while keeping the latest meta data even if less than tillTime.
-   * @param tillTime time till the meta data will be deleted.
-   * @param callback callback called before deleting a meta data column.
-   * @return total number of columns deleted.
+   * Remove metadata for provided list of log files
+   *
+   * @param filesToRemove Map of files to be removed to namespace
+   * @return count of removed meta files
    */
-  public int cleanMetaData(final long tillTime, final DeleteCallback callback) throws Exception {
+  public int cleanMetadata(final Map<URI, NamespaceId> filesToRemove) {
     return execute(new TransactionExecutor.Function<Table, Integer>() {
```

Comment:
do you have to scan the table for this? Can't you just do a batch delete of all rows/columns in filesToRemove?